stacks-core/src/address.rs

```diff
@@ -0,0 +1,253 @@
+use std::fmt;
+
+use crate::{
+    c32::{decode_address, encode_address},
+    crypto::{
+        hash::{Hash160, SHA256Hash},
+        PublicKey,
+    },
+    StacksError, StacksResult,
+};
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum AddressHashMode {
+    SerializeP2PKH,
+    SerializeP2SH,
+    SerializeP2WPKH,
+    SerializeP2WSH,
+}
+
+#[derive(Debug, Clone)]
+pub struct StacksAddress {
+    version: u8,
+    hash: Hash160,
+}
+
+impl StacksAddress {
+    pub fn new(version: u8, hash: Hash160) -> Self {
+        Self { version, hash }
+    }
+
+    pub fn version(&self) -> u8 {
+        self.version
+    }
+
+    pub fn hash(&self) -> &Hash160 {
+        &self.hash
+    }
+
+    pub fn from_public_keys(
+        version: u8,
+        public_keys: &[PublicKey],
+        signatures: usize,
+        hash_mode: AddressHashMode,
```

Comment:
I have always found these args slightly confusing. It would be great if we could find a more intuitive interface here. I'm not sure what that would be though, but especially the `signatures` arg is confusing to me. Could we at least have it documented?

Reply:
Ah I didn't even see this comment when I wrote https://github.com/stacks-network/sbtc/issues/7

I have a lot of ideas on how to improve it.